pennylane/devices/qubit/simulate.py

```diff
@@ -237,7 +257,297 @@ def simulate(
+    idx_sample = find_measurement_values(circuit)
+    for i in reversed(idx_sample):
+        aux_circuit._measurements.pop(i)
+    for i, m in enumerate(circuit.measurements):
+        if isinstance(m, VarianceMP) and m.mv is None:
+            aux_circuit._measurements[i] = SampleMP(obs=m.obs)
+    return aux_circuit
+
+
+def simulate_one_shot_native_mcm(
+    circuit: qml.tape.QuantumScript,
+    rng=None,
+    prng_key=None,
+    debugger=None,
+    interface=None,
+) -> Result:
+    """Simulate a single shot of a single quantum script with native mid-circuit measurements.
+
+    Args:
+        circuit (QuantumTape): The single circuit to simulate
+        rng (Union[None, int, array_like[int], SeedSequence, BitGenerator, Generator]): A
+            seed-like parameter matching that of ``seed`` for ``numpy.random.default_rng``.
+            If no value is provided, a default RNG will be used.
+        prng_key (Optional[jax.random.PRNGKey]): An optional ``jax.random.PRNGKey``. This is
+            the key to the JAX pseudo random number generator. If None, a random key will be
+            generated. Only for simulation using JAX.
+        debugger (_Debugger): The debugger to use
+        interface (str): The machine learning interface to create the initial state with
+
+    Returns:
+        tuple(TensorLike): The results of the simulation
+        dict: The mid-circuit measurement results of the simulation
+    """
+    state, is_state_batched, mcm_dict = get_final_state(
+        circuit, debugger=debugger, interface=interface
+    )
+    return (
+        measure_final_state(circuit, state, is_state_batched, rng=rng, prng_key=prng_key),
+        mcm_dict,
+    )
+
+
+def accumulate_native_mcm(circuit: qml.tape.QuantumScript, all_shot_meas, one_shot_meas):
+    """Incorporates new measurements in current measurement sequence.
+
+    Args:
+        circuit (QuantumTape): A one-shot (auxiliary) QuantumScript
+        all_shot_meas (Sequence[Any]): List of accumulated measurement results
+        one_shot_meas (Sequence[Any]): List of measurement results
+
+    Returns:
+        tuple(TensorLike): The results of the simulation
+    """
+    new_shot_meas = [None] * len(circuit.measurements)
+    if not isinstance(all_shot_meas, (list, tuple)):
+        return accumulate_native_mcm(circuit, [all_shot_meas], one_shot_meas)
+    if not isinstance(one_shot_meas, (list, tuple)):
+        return accumulate_native_mcm(circuit, all_shot_meas, [one_shot_meas])
+    for i, m in enumerate(circuit.measurements):
+        if isinstance(m, CountsMP):
+            tmp = Counter(all_shot_meas[i])
+            tmp.update(Counter(one_shot_meas[i]))
+            new_shot_meas[i] = tmp
+        elif isinstance(m, SampleMP):
+            if not isinstance(all_shot_meas[i], (list, tuple)):
+                new_shot_meas[i] = [all_shot_meas[i]]
+            else:
+                new_shot_meas[i] = all_shot_meas[i]
```

Comment:
Following on my comment up above.  Could we standardize the shape and requirements of an `all_shot_meas` type so that the sample entry is always a list?

Reply:
On first pass the sample comes out as in an `ndarray`. I then accumulate all the arrays in a list for simplicity. We could preallocate an array and truncate it depending on the number of thrown-away shots with post-selection, but I didn't do that here since the feature isn't part of the PR.